[2단계 - 자동차 경주] 기린(정유정) 미션 제출합니다.

jeongyou opened the pull request:
안녕하세요. 2단계 리뷰도 잘 부탁드립니다!

## 리팩터링 요구 사항

- 도메인 로직은 domain/ 하위로, UI 관련 로직은 view/ 하위에서 관리한다.
- domain/ 하위의 모듈은 view/ 하위의 모듈을 의존하지 않아야 한다.
- 도메인 로직 내에서 테스트하기 어려운 부분을 분리하고, 테스트 가능한 부분에 대해서만 단위 테스트를 작성한다.
- 테스트 코드에서 jest.fn()을 사용하지 않는다. (마찬가지로, mock/spy 등도 이번 미션에서는 사용하지 않는다.)

## 리팩터링 내용

**1. InputView, OutView, Validate를 클래스에서 함수 모듈로 변경하였습니다.**
  - 세 클래스 모두 따로 필드를 저장할 일이 없다고 생각하였습니다. 그래서 불필요한 객체 생성을 없애기 위해 단순하게 메서드들만 정의한 모듈로 만들어 깔끔하게 코드를 작성하였습니다.

    #### ❓궁금한 점
    - 그렇다면 RacingCar 클래스도 현재 메서드들로만 이루어져있는데 이런 경우에도 함수 모듈로 만드는게 맞는지 궁금합니다.

**2. roundOfRacing의 불변성을 위해 roundOfRacing에서 원본 cars 배열을 직접 수정하지 않고, 새로운 raceResult 배열을 반환하도록 변경하였습니다.**
- 이전의 피드백에서 `cars 배열은 roundOfRacing 메서드의 호출으로 인해 영구적으로 수정되며, 이전의 상태를 알기 어려워집니다.` 라고 말씀하셔서 반영 해보았습니다.

  ```javascript
      oneRound(cars) {
          cars.forEach(car => {
              car.move(getRandomNumber(0, 10));
          });
          return cars;
      }

      #roundOfRacing(racingCar, cars, attempts) {
      OutputView.printResultMessage();
      let raceResult = [...cars];

      for (let i = 0; i < attempts; i++) {
        raceResult = racingCar.oneRound(raceResult);
        OutputView.printOneRoundResult(raceResult);
      }

      return raceResult;
    }
  ```
    #### ❓궁금한 점
    - 하지만 결국 onRound() 함수에서 car의 상태가 변하게 되어 불변성이 지켜지는지 모르겠습니다. 만약 새로운 객체를 만들어서 반환한다면 이전의 경주 기록이 모두 사라지게 되어 car 객체를 어쩔 수 없이 변경하게 되는데 해결하기 위한 다른 방법이 있는지 궁금합니다.

**3. domain과 View를 분리하였습니다.**
- domain 폴더에는 `Car.js`, `RacingCar.js`, `Validate.js`를, view 폴더에는 `InputView.js`, `OutputView.js`를 넣어 구분하였습니다.
- *domain/ 하위의 모듈은 view/ 하위의 모듈을 의존하지 않아야 한다.*는 리펙터링 요구사항에 따라 도메인과 뷰를 연결하는 역할을 `RacingCarManager.js`에서 `App.js`로 변경하였습니다.
  #### ❓궁금한 점
  - 역할에 따라 domain과 view가 잘 분리 되었는지 궁금합니다.
  - 같은 domain 끼리는 서로 import 해도 되는것인지 궁금합니다. 

**4. 상수를 별도의 파일로 분리하였습니다.**
- 상수가 많아지는 경우를 생각해 별도의 파일로 관리하는 것이 좋은 방법이라고 판단해 상수를 별도의 파일로 분리하였습니다.

#### ❓궁금한 점
  - 상수를 분리할 때 view에서 사용하는 상수랑 domain에서 사용하는 상수를 구분하여 각 폴더에 넣는 것이 효율적이라고 생각해서 각 폴더에 상수파일을 하나씩 만들었는데 보통은 하나의 파일에서 관리를 하는지, 아니면 어떤 다른 방법들이 있는지 궁금합니다.

**5. 랜덤 숫자 생성 함수를 util로 분리하였습니다.**
- 랜덤 숫자 생성 함수를 재사용 가능한 형태로 수정한 후 분리하였습니다.
#### ❓궁금한 점
  - 저는 랜덤 숫자 생성 함수를 min, max를 입력 할 수 있도록 변경해야 재사용 가능하다고 판단하여 함수를 수정했는데 원래의 0에서 9까지의 랜덤 숫자를 생성해주도록 고정되어있는 함수같은 경우도 util로 분리하는지 궁금합니다.

**6. readLineAsync를 util로 분리하였습니다.**
- 원래 InputView에서만 사용하는 함수이기 때문에 같은 파일에 작성했었는데 다른 곳에서도 그대로 사용 가능하다고 생각해서 util로 분리 하였습니다.

**7. 폴더 구조를 domain, util, view로 분리하여 정리하였습니다.**
  ```
  src/
  ├─ domain/ // 도메인 로직
  │  ├─ Car.js // 자동차 클래스
  │  ├─ constants.js // 도메인에서 사용하는 상수
  │  ├─ RacingCar.js // 자동차 경주 로직을 담당하는 클래스
  │  └─ Validate.js // 검증 함수 모듈
  ├─ util/ // 재사용 가능한 함수를 모아둠
  │  ├─ random.js // 랜덤 숫자 생성 함수
  │  └─ readLineAsync.js // 입력 유틸 함수
  ├─ view/ // UI 관련 로직
  │  ├─ constants.js // 뷰에서 사용하는 상수
  │  ├─ InputView.js // 입력 함수 모듈
  │  └─ OutputView.js // 출력 함수 모듈
  ├─ App.js // 도메인과 뷰를 연결 하는 역할
  └─ index.js
  ```

  #### ❓궁금한 점
  - 잘 분리가 된 구조인지, 아니라면 더 좋은 방법이 있는지 궁금합니다.


## 이외의 질문
1. RacingCar.js

```javascript
oneRound(cars) {
        cars.forEach(car => {
            car.move(getRandomNumber(0, 10));
        });
        return cars;
    }
```

RacingCar에 있는 함수중에 onRound() 함수만 랜덤 값을 지정해 줄 수 없어서 테스트 하지 못했습니다. 하지만 car.move()를 테스트 했기 때문에 이 함수 또한 자동으로 테스트 된거나 다름 없다는 생각이 드는데 이 함수에 대해서도 테스트 할 수 있도록 리펙터링 하는것이 좋은지 궁금합니다.

감사합니다.

eunsoA committed:
docs : 기능구현목록 작성

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
docs : 기능구현목록 작성

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
docs: 기능 구현 목록 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
docs: 기능 구현 목록 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
feat: 자동차 이름을 입력 받은 후 자동차 생성

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
test: 자동차 이름을 입력받은 후 자동차 생성 에 대한 예외 사항 테스트

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
test: 자동차 이름을 입력받은 후 자동차 생성에 대한 테스트

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
refactor: 자동차 이름 명령문 문구 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
feat: 시도할 횟수를 입력 받는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
test: 시도 횟수 입력에 대한 예외 사항 테스트

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
feat: 각 자동차의 전진 조건 구하는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
test: 각 자동차의 전진 조건에 대한 단위 테스트

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
docs: 기능 구현 목록 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
feat: 한 회차 실행 결과를 출력하는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
docs: 기능 구현 목록 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
feat: 입력받은 시도 횟수 만큼 3번, 4번 반복 기능 추가

- 3번: 각 자동차의 전진 조건 구하기
- 4번: 한 회차 실행 결과 출력

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
feat: 가장 멀리간 자동차 구하는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
feat: 최종 우승자 출력하는 기능 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
fix: 가장 멀리간 position 계산 오류 수정

- getMaxPosition 함수 기능 수정

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
test: 이름을 가진 자동차를 만드는 테스트 수정

- toBeDefined()에서 toEqual()로 변경

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
test: RacingManager Class 테스트

- createCars() 와 getWinners() 기능 테스트

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
refactor: 에러 메세지 출력 형식 변경

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
refactor: 하드코딩된 값을 상수로 변경

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

eunsoA committed:
style: EOL 개행 추가

Co-authored-by: Youjeong Jeong <jeongyou@users.noreply.github.com>

jeongyou committed:
Merge branch 'jeongyou' of https://github.com/jeongyou/javascript-racingcar into jeongyou

jeongyou committed:
refactor: OutputView 클래스를 함수 기반으로 변경

- OutputView 클래스 없이 함수 기반으로 출력 기능을 수행하도록 변경
- 불필요한 클래스 제거로 코드 구조 간소화

jeongyou committed:
refactor: InputView 클래스를 함수 기반으로 변경

- InputView 클래스 없이 함수 기반으로 출력 기능을 수행하도록 변경
- 불필요한 클래스 제거로 코드 구조 간소화

jeongyou committed:
refactor: InputView, OutputView 함수 기반 변경에 따른 RacingcarManager 수정

InputView와 OutputView를 클래스에서 함수 기반으로 변경함에 따라 RacingcarManager 내부 import 방식 및 호출 방식을 수정
- InputView, OutputView의 메서드 호출을 객체 참조 방식에서 개별 함수 호출 방식으로 변경
- 필요 없는 클래스 인스턴스화 제거
- 코드 구조 단순화 및 가독성 개선

jeongyou committed:
refactor: roundOfRacing의 불변성 유지 및 oneRound 리팩토링

roundOfRacing에서 원본 cars 배열을 직접 수정하지 않고, 새로운 raceResult 배열을 반환하도록 변경함.
- 기존 cars 배열을 유지하고 새로운 raceResult를 반환하여 불변성을 유지
- oneRound에서 map()을 사용하여 새로운 Car 객체를 반환하도록 수정

jeongyou committed:
refactor: 도메인 로직과 뷰 로직 분리 및 구조 개선

- RacingcarManager에서 InputView, OutputView, Validate 의존성 제거
- App.js에서 입력/출력 로직 및 유효성 검증 처리하도록 변경
- domain 폴더와 view 폴더로 파일 이동하여 책임 분리
- 불필요한 의존성 제거

jeongyou committed:
refactor: 뷰 관련 상수를 별도 파일로 분리

- UI에서 사용하는 상수를 src/view/constants.js로 이동
- 기존 코드에서 상수를 import하도록 수정

jeongyou committed:
refactor: 도메인 관련 상수를 별도 파일로 분리

- 도메인에서 사용하는 상수를 src/domain/constants.js 파일로 이동
- 기존 코드에서 상수를 import하도록 수정

jeongyou committed:
refactor: 폴더 구조 변경에 따른 import 경로 수정 및 불필요한 import 문 제거

- 폴더 구조 변경에 맞게 import 경로 업데이트
- 사용하지 않는 import 문 삭제하여 코드 정리

jeongyou committed:
refactor: 랜덤 숫자 생성 함수 분리 및 RacingcarManager 수정

- 기존 RacingcarManager 내부의 getRandomNumber() 제거
- 랜덤 숫자 생성 함수를 util.js로 분리
- RacingcarManager에서 getRandomNumber()을 import하여 사용

jeongyou committed:
refactor: 랜덤 숫자 생성 함수를 util/random.js로 분리

- 기존 util 파일에 있던 getRandomNumber 함수를 util/random.js로 이동
- 관련 모듈에서 import 경로 업데이트

jeongyou committed:
refactor: readLineAsync 함수를 src/util/readLineAsync.js로 분리

- InputView 내에 있던 readLineAsync 함수를 별도 파일로 이동하여, 입출력 관련 유틸리티를 분리
- 관련 모듈에서 import 경로 업데이트

jeongyou committed:
refactor: Validate 클래스를 함수 기반 모듈료 변경

- 기존 Validate 클래스를 개별 검증 함수 모음으로 변경

jeongyou committed:
test: Validate 테스트 코드 업데이트

- Validate 클래스를 함수 기반 모듈로 전환한 변경 사항에 맞게 테스트 코드 수정

jeongyou committed:
refactor: Validate를 함수 호출 방식으로 수정

jeongyou committed:
refactor: 파일명을 RacingCar.js로 변경

swon3210 commented:
안녕하세요 기린! 리뷰가 늦어져서 죄송합니다! 최근 야근이 잦아져서 리뷰를 진행할 시간을 확보하는데 어려움을 느끼고 있는데요. 기린의 과제 일정에 차질이 없도록 적어도 리뷰가 늦어질 것 같으면 언제 리뷰가 완료될 수 있는지 미리 말씀드리도록 하겠습니다.

swon3210 commented:
[ 개별 질문에 대한 답변 ]

>세 클래스 모두 따로 필드를 저장할 일이 없다고 생각하였습니다. 그래서 불필요한 객체 생성을 없애기 위해 단순하게 메서드들만 정의한 모듈로 만들어 깔끔하게 코드를 작성하였습니다. 그렇다면 RacingCar 클래스도 현재 메서드들로만 이루어져있는데 이런 경우에도 함수 모듈로 만드는게 맞는지 궁금합니다.

- 맞다고 말씀드리기 보다는 '적합하다'고 말씀드리는 것이 좋을 듯합니다. RacingCar 클래스는 프로퍼티가 없기에 개별적인 상태를 지니고 있지 않습니다. 매개변수를 받아서 연산을 대신 실행해줄 뿐입니다. RacingCar 를 클래스로 두더라도 코드를 이해하는데는 전혀 문제가 없지만, RacingCar 라는 개념을 통해 각각의 메서드를 응집해줄 이유는 없어보입니다. 

- 현업에서는 하나의 도메인에서 사용되는 로직이 다른 도메인에서도 사용되어야 한다면 이를 위해 메서드를 정의하는 위치를 바꾸는 경우가 많습니다. 코드를 옮기기 편한 형태는 각각을 함수로 빼둔다면 이러한 작업이 좀 더 수월해지는 장점이 있습니다.



>하지만 결국 onRound() 함수에서 car의 상태가 변하게 되어 불변성이 지켜지는지 모르겠습니다. 만약 새로운 객체를 만들어서 반환한다면 이전의 경주 기록이 모두 사라지게 되어 car 객체를 어쩔 수 없이 변경하게 되는데 해결하기 위한 다른 방법이 있는지 궁금합니다.

- 맞습니다. oneRound 함수는 cars 배열 안의 car 객체 각각의 상태를 변경하기에 엄밀한 의미에서 불변성을 지키지 못합니다. 상태의 불변성을 엄밀하게 지키기 위해서는 기존 car 객체의 상태는 그대로 두고, 업데이트가 반영된 새로운 car 객체들의 배열을 반환하도록 oneRound 함수를 수정할 필요가 있습니다. 예를들어 아래와 같이 할 수 있습니다.

```javascript
  oneRound(cars) {
    return cars.map((car) => {
      const newCar = new Car(car.name);
      newCar.position = car.position;

      if (getRandomNumber(0, 10) >= 4) {
        newCar.position += 1;
      }

      return newCar;
    });
  }
```
위의 코드가 맘에 드시지는 않으리라 생각합니다. 제일 좋은 방법은 car 객체를 단순 자바스크립트 Object 로서 다루도록 구조를 변경하는 것이겠지만 그건 코드 구조를 너무 많이 바꿔야 합니다.

지금의 oneRound 함수로도 요구사항을 충족시키고, 이해하기 쉬운 코드를 작성하는데 문제가 없습니다.


>역할에 따라 domain과 view가 잘 분리 되었는지 궁금합니다.
>같은 domain 끼리는 서로 import 해도 되는것인지 궁금합니다.

- domain 은 잘 분리되었다고 생각합니다. 다만 validate 메서드 중 'carNameLength' 와 같은 메서드들은 이름에서 알 수 있듯이 car 도메인에 종속된 메서드입니다. 가능하면 car 도메인과 같이 응집시키는 것이 좋습니다.

- view 는 특정 도메인에 종속된 부분들이 보입니다. 예를들어 'printOneRoundResult' 메서드는 cars 배열을 받도록 되어 있습니다. 해당 메서드는 cars 배열 외의 다른 매개변수를 받을 수 없을 것이기에, 다른 도메인에서는 재사용이 제한적일 수밖에 없습니다. 

- 어떤 함수가 특정 도메인에 종속적인지의 여부를 따져볼 때는 특정 도메인의 수정사항이 해당 함수의 수정으로 이어지는 지를 생각해보면 좋습니다. 예시로 든 'printOneRoundResult' 메서드는 car 객체의 형식이 바뀌면 함수 내부 로직도 같이 바꾸어야 하도록 되어 있습니다.

- 종속성을 제거하기 위해 cars 가 아닌 도메인과 독립적인 매개변수를 받도록 처리할 수 있습니다. 다음과 같은 형식으로 바꿔보는 것을 생각해보시면 좋습니다.



>상수를 분리할 때 view에서 사용하는 상수랑 domain에서 사용하는 상수를 구분하여 각 폴더에 넣는 것이 효율적이라고 생각해서 각 폴더에 상수파일을 하나씩 만들었는데 보통은 하나의 파일에서 관리를 하는지, 아니면 어떤 다른 방법들이 있는지 궁금합니다.

- 작은 프로젝트에서는 하나의 파일에서 모든 상수를 관리할 수 있겠지만 대부분의 경우, 기린이 해주신 방식이 훨씬 좋은 방식이라고 생각합니다. 관련 있는 코드들을 응집시키게 되면 특정 플로우를 이해하는데 읽어야 하는 코드의 범주가 한정되어 전반적인 유지보수 효율이 올라가는데요. 상수 파일 또한 사용처와 최대한 가까운 곳에 따로 위치시키는 것이 좋다고 생각합니다.


>저는 랜덤 숫자 생성 함수를 min, max를 입력 할 수 있도록 변경해야 재사용 가능하다고 판단하여 함수를 수정했는데 원래의 0에서 9까지의 랜덤 숫자를 생성해주도록 고정되어있는 함수같은 경우도 util로 분리하는지 궁금합니다.

- util 은 코드 작성의 편의성을 높이기 위한 함수를 의미합니다. 따라서 호출에 필요한 조건이 단순할 수록 좋습니다. min max 를 유동적으로 설정하는 것이 요구사항에 필수적이라면 기린처럼 작성하는 것이 맞습니다. 그러나 이번 과제처럼 딱히 그럴 필요가 없다면, 0에서 9까지의 랜덤 숫자를 생성하는 고정된 역할을 수행하는 함수도 util 로써 둘 수 있을 것입니다. 물론 동작을 더 잘 설명하기 위해 generateZeroToNineNumber 와 같은 식으로 min max 매개변수를 없애는 대신 함수명을 구체화 해볼 수도 있을 것입니다.


>잘 분리가 된 구조인지, 아니라면 더 좋은 방법이 있는지 궁금합니다.

- 매우 잘 분리된 구조입니다. 각각이 서로의 영역을 침범하지 않는 구조를 만들어주셨습니다.

>RacingCar에 있는 함수중에 onRound() 함수만 랜덤 값을 지정해 줄 수 없어서 테스트 하지 못했습니다. 하지만 car.move()를 테스트 했기 때문에 이 함수 또한 자동으로 테스트 된거나 다름 없다는 생각이 드는데 이 함수에 대해서도 테스트 할 수 있도록 리펙터링 하는것이 좋은지 궁금합니다.

- oneRound 함수는 출력이 일정하지 않은 함수이기에 테스트하기 어려운 함수입니다. 어떻게 리팩터링 하더라도 테스트 하기 어렵기는 매한가지일 것입니다. 이 경우엔 해당 함수는 테스트 하지 않는 것이 좋은 선택일 수 있습니다.

- oneRound 함수를 매우 많은 횟수로 호출시켜서 시행 결과의 평균이 확률적 기대값과 유사한지를 확인하는 괴랄(?)한 방법도 있습니다. 시행 횟수가 매우 크게 늘어날수록, 결과값의 분포는 정규 분포를 따르기 때문입니다. 하지만 해당 함수의 동작이 매우 단순하다는 것을 생각해보면 그 정도의 리소스를 투자해서 테스트할 필요는 없어보입니다.

swon3210 commented on src/App.js:
```diff
@@ -1,10 +1,85 @@
   async run() {
-    const racingcarManager = new RacingcarManager();
-    await racingcarManager.start();
+    const racingCar = new RacingCar();
+
+    const carNames = await this.#getCarNames();
+    const attempts = await this.#getAttempts();
+
+    const cars = racingCar.createCars(carNames);
+
+    const raceResult = this.#roundOfRacing(racingCar, cars, attempts);
+    const winners = racingCar.getWinners(raceResult);
```
전체 어플리케이션의 동작 플로우가 간결하면서도 한눈에 보이도록 비즈니스 로직을 잘 작성해주셨습니다!

jeongyou committed:
refactor: 클래스 구조를 함수 모듈로 변경

- 기존 RacingCar 클래스를 객체를 반환하는 함수 모듈로 변경

jeongyou committed:
test: Racing 함수 모듈 구조에 맞게 테스트 코드 수정

- import RacingCar from './RacingCar.js'에서 new RacingCar() 방식 제거
- 함수 호출 방식 수정 (RacingCar.createCars() 형태로 변경)

jeongyou committed:
refactor: 컨트롤러에서 RacingCar 함수 모듈을 불러와 사용하도록 수정

- import RacingCar from './RacingCar.js' 이후 객체 생성 코드 제거
- RacingCar.createCars(carNames) 같은 함수 호출 방식으로 변경

jeongyou committed:
refactor: printOneRoundResult의 도메인(Car 객체) 종속성 제거

- printOneRoundResult 함수가 특정 도메인(Car 객체 배열)에 의존하지 않도록 수정
- 기존에는 Car 객체 배열(cars)을 직접 받아서 처리했지만, 일반적인 데이터 배열({ name, position })을 받도록 변경하여 도메인과 독립적인 출력이 가능하도록 개선

jeongyou commented:
크리스 안녕하세요! 바쁜 와중에 꼼꼼하게 답변 해주신 내용들이 많은 도움이 되었습니다!

말씀해 주신 내용을 바탕으로 코드를 조금 더 리팩토링해보았습니다. 

1. RacingCar 클래스를 함수 모듈로 변경   [88f0eaf](https://github.com/woowacourse/javascript-racingcar/pull/409/commits/88f0eaf13410fcf2bc0172b62fc2546703223d86)
    - RacingCar 클래스는 프로퍼티를 가지지 않으므로, 굳이 클래스로 유지할 필요가 없다고 판단하여 함수 모듈로 리팩토링하였습니다.

2. printOneRoundResult의 도메인(Car 객체) 종속성 제거   [c3cff58](https://github.com/woowacourse/javascript-racingcar/pull/409/commits/c3cff58cdc216d997550cb8d11385ab3f73ed60d)
    - 질문에 답변 주신대로 기존 printOneRoundResult는 Car 객체의 구조가 변경될 경우 함수 내부 로직도 함께 수정해야 하는 종속적인 구조였습니다. 그래서 이를 개선하기 위해 Car 객체에 직접 의존하지 않고, { name, position } 형태의 일반 데이터 배열을 받도록 변경하여 Car 객체와의 종속성을 제거하였습니다.

혹시 잘못된 부분이 있다면 말씀해주시면 감사하겠습니다!

swon3210 reviewed (approved):
이번 과제처럼 요구사항이 단순한 서비스는 UI 와 도메인의 완전한 분리의 필요성을 체감하기가 어려웠을텐데 피드백을 빠르게 이해하고 반영해주셨네요.

중요한 변경 사항을 커밋 해시를 통해 전달 주시는 것도 매우 중요한 소프트 스킬이라고 생각합니다. 은근히 귀찮은 부분이라 잘 챙기는 사람은 많이 없는 것 같습니다.

미션 진행하시느라 수고하셨습니다 기린! 다음 미션도 화이팅입니다!
